### PR TITLE
falco-driver-loader - break apart logic and improvements

### LINKED ISSUE #1125: falco-driver-loader improvements proposal
**Motivation**

As part of the ongoing discussions about the [scope of Falco artifacts](https://github.com/falcosecurity/falco/issues/1114) we need an officially supported tool for building, downloading, and installing various drivers. Currently, the `falco-driver-loader` takes care of that, and it is already included in the new [stable binary distribution](https://dl.bintray.com/falcosecurity/bin/x86_64/). 

Furthermore, the source of `falco-driver-loader` needs to stay into the Falco repo,  so it can easily be kept in sync with Falco and the right parameters can be [substituted into the script](https://github.com/falcosecurity/falco/blob/master/scripts/falco-driver-loader#L387) during the building process. Also, having it in the same place helps with maintenance.

That being said, there are other tools that need to perform similar actions: for example [falcoctl installs the kernel module](https://github.com/falcosecurity/falcoctl/blob/master/pkg/kernelmoduleloader/kernel_module_loader.go) by replicating some of the logic already included in `falco-driver-loader`. Moreover, the installation logic can change over time and we will need a way to version this process too.

For these reasons, I would like to propose several improvements in order to make it usable by other tools and avoid duplicating the same logic in other places.

**Feature**

- [ ] Make explicit that `falco-driver-loader` is versioned alongside Falco (mostly a documentation issue)
- [ ] Do not unload the driver if is already present and the version matches (needed to avoid problems with manually loaded drivers)
- [ ] ~~Distribute another artifact that includes just `falco-driver-loader` so it can be quickly downloaded and used by both users and other tools (ie. `falcoctl`), the artifact will be versioned as same as Falco~~ :point_right: https://github.com/falcosecurity/falco/issues/1159
- [ ] Add a simple interface that allows executing single actions (ie. only download, only install, or only build) so it can be easily used by other tools (ie. `falcoctl`)
- [x] Update `falco-driver-loader` in order to download artifact from https://bintray.com/falcosecurity

**Alternatives**

**Additional context**

### LINKED ISSUE #1193: Break apart logic in `falco-driver-loader` or rename the script
**Motivation**

Right now our documentation and our calls use the term `loader` to define the behavior of [this script](https://github.com/falcosecurity/falco/blob/master/scripts/falco-driver-loader).

I think that the name could be improved to clarify exactly what the script is doing.

I also believe that the script should be dynamic and configurable to perform single tasks without making assumptions on behalf of the user. 

**Feature**

Right now the script performs multiple phases of logic within monolithic functions that perform more than one task.

For instance `load_kernel_module` does more than just `load` something. It also tries to download a kernel module, it also tries to compile a kernel module. 

I think we should take one of 2 paths

1) Rename the script to something that describes what it does.

```bash
./falcodriver-loader-downloader-compiler
```

2) Fix the script so it is modular and can be used as a single use case

```bash
# This will try to download first, and then attempt to compile if the download fails
./falcodriver module --download --compile 

# This will never download, and will only try to compile
./falcodriver bpf --compile

# This will first try to use DKMS to load the kernel module if it exists on the system, then download, then compile
./falcodriver module --load --download --compile

# This will do exactly what it does now and try to detect and succeed at all costs
./falcodriver # no arguments
```

**Alternatives**

**Additional context**

### PR DESCRIPTION
**Pull request by leogr**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area engine

> /area examples

> /area rules

> /area integrations

> /area tests

> /area proposals

**What this PR does / why we need it**:
See #1193

**Which issue(s) this PR fixes**:

Fixes #1193
Fixes #1125 (again)

**Special notes for your reviewer**:

```shell
Usage:
  falco-driver-loader [driver] [options]

Available drivers:
  module        kernel module (default)
  bpf           eBPF probe

Options:
  --help        show brief help
  --compile     try to compile the driver locally
  --download    try to download a prebuilt driver
```

With no options and arguments, it preserves the previous behavior (we should test it deeply).

Notable improvements in messages:
 - All info messages start with `* Message...`
 - Success messages with `* Success: message...`
 - Messages without `*` represent the output of another program or an error
 - Verbosity increased 

/cc @kris-nova 
/cc @leodido 
/cc @fntlnz 

### falco-driver-loader (pre refactoring behavior)

Preflight check:
- if not root, then exit 1
- if not `curl`, then exit 1

#### module

Preflight check:
- if not `lsmod`, then exit 1
- if not `modprobe`, then exit 1
- if not `rmmod`, then exit 1

1. Unload module, if present
2. If module still loaded, then exit 0
3. If UEK host, then
    - do nothing
    *else*
    - if `dkms install` and `insmod` then exit 0
4. if `modprobe` module, then exit 0
5. if unsupported OS, then exit 1 (see `get_target_id`)
6. if `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.ko`, then:
    - if `insmod` then exit 0, else exit 1
7. if `curl` module, then
    - if `insmod` then exit 0, else exit 1
    *else*
    - exit 1
8. exit 0

#### bpf 

Preflight check:
- Mount debugfs (continue on error)
- if kernel config not found, exit 1 (see `get_kernel_config`)
- if unsupported OS, then exit 1 (see `get_target_id`)

1. if NOT `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, then go to step 6.
    - if COS, MINKUBE, or `BPF_USE_LOCAL_KERNEL_SOURCES` then download kernel source
        - if download failed, then exit 1
        - `customize_kernel_build`
    - make bpf probe to `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`
2. if NOT `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, 
    - `curl` bpf probe (continue on error)
3. if `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, then
    - if symlink to `~/.falco/falco-bpf.o` then exit 0, else exit 1
    *else*
    - exit 1
4. exit 0

**N.B.**: **bpf** exists with 0 even though `curl` failed (see step 2.)

**Does this PR introduce a user-facing change?**:

```release-note
update(scripts): improve `falco-driver-loader` output messages
new(scripts): options and command-line usage for `falco-driver-loader`
```

### CODE REVIEW

File: scripts/falco-driver-loader [hunk 1]
```diff
@@ -420,18 +480,100 @@ fi
...
+	ENABLE_DOWNLOAD="yes"
+fi
+
 if [ "$(id -u)" != 0 ]; then
-	echo "Installer must be run as root (or with sudo)."
+	>&2 echo "This program must be run as root (or with sudo)"
 	exit 1
 fi
 
 if ! hash curl > /dev/null 2>&1; then
```
Comment (leodido) [inline scripts/falco-driver-loader:566]: Can you please add a check for `dkms` when user wants to compile?

Comment (leogr) [reply scripts/falco-driver-loader:566]: It is already present [here](https://github.com/falcosecurity/falco/pull/1200/files#diff-b1a242527066116a4e7485802c03bbb3R154) (see the `else` branch).
The check will not terminate the execution as like as the previous script version.

File: scripts/falco-driver-loader [hunk 2]
```diff
@@ -140,6 +139,84 @@ get_target_id() {
...
+					echo "* Running dkms build failed, couldn't find ${DKMS_LOG}"
+				fi
+			fi
+		else
+			echo "* Skipping dkms install (dkms not found)"
+		fi
+	fi
+
+	if [ -n "$SKIP_LOAD" ]; then
+		if modprobe "${DRIVER_NAME}" > /dev/null 2>&1; then
```
Comment (leogr) [inline scripts/falco-driver-loader:186]: It should be a `--dry-run`

File: scripts/falco-driver-loader (hunk 1 above)
Comment (leogr) [reply scripts/falco-driver-loader:566]: Furthermore, we need to continue the execution so the download strategy can happen.
When `dkms` does not exist, users will see `* Skipping dkms install (dkms not found)`

File: scripts/falco-driver-loader [hunk 3]
```diff
@@ -420,18 +480,100 @@ fi
...
+      	ENABLE_COMPILE="yes"
+	    has_opts="true"
+      shift
+      ;;
+    --download)
+	    ENABLE_DOWNLOAD="yes"
+	    has_opts="true"
+      shift
+      ;;
+	--skip-load)
```
Comment (leodido) [inline scripts/falco-driver-loader:539]: Indentation

File: scripts/falco-driver-loader [hunk 4]
```diff
@@ -420,18 +480,100 @@ fi
...
+	    has_args="true"
+      shift
+    fi
+      ;;
+    -h|--help)
+      print_usage
+      exit 0
+      ;;
+    --compile)
+      	ENABLE_COMPILE="yes"
```
Comment (leodido) [inline scripts/falco-driver-loader:530]: Indentation

File: scripts/falco-driver-loader [hunk 5]
```diff
@@ -177,226 +241,222 @@ load_kernel_module() {
...
+			load_bpf_probe_compile
+		fi
+	fi
 
-		curl -L --create-dirs "${FALCO_DRIVER_CURL_OPTIONS}" -o "${HOME}/.falco/${BPF_PROBE_FILENAME}" "${URL}"
+	if [ -n "$ENABLE_DOWNLOAD" ]; then
+		if [ -f "${HOME}/.falco/${BPF_PROBE_FILENAME}" ]; then
+			echo "* Skipping download, eBPF probe is already present in ${HOME}/.falco/${BPF_PROBE_FILENAME}"
+		else
+			load_bpf_probe_download
```
Comment (leodido) [inline scripts/falco-driver-loader:440]: Check the operation was successful so to give hints to the user

**Comment by leogr** [comment]
> The only thing I'm not really sure about is the usage of `check_skip_load` to circuit-break only the load phase of the drivers (not their installation).
> 
> So, implemented in this way, what's the use case of `--skip-load`? Maybe I'm missing something..

My thought when I saw the proposal in the issue was... we might need this for testing purposes only. Ofc, Falco will not work if `--skip-load` is used.

Should we remove it?

**Comment by leogr** [comment]
Hey @fntlnz 
I'm working on tests, so we can add them.

That being said, really important design question: after talking with @leodido I believe we should remove the `--skip-load` option. WDYT?
If you agree on to remove it too, I will remove it ASAP.

**Comment by leodido** [comment]
I think `--skip-load` flag should be removed: it adds more complexity than it removes. 

The `--download` and `--compile` flags are pretty explanatory and they cover the use cases

My 2 cents

File: scripts/falco-driver-loader (hunk 5 above)
Comment (leogr) [reply scripts/falco-driver-loader:424]: Fixed here :point_right: https://github.com/falcosecurity/falco/pull/1200/files#diff-b1a242527066116a4e7485802c03bbb3R378

**Comment by leogr** [comment]
Update:
- `--skip-load` has been removed
- do not continue when epbf download fails

**Comment by leogr** [comment]
**IMPORTANT**

I have noticed that:
```
	echo "* Trying to load a system ${DRIVER_NAME} driver, if present"
	if modprobe "${DRIVER_NAME}" > /dev/null 2>&1; then
		echo "* Success: ${DRIVER_NAME} module found and loaded with modprobe"
		exit 0
	fi
```
might load a kernel module that does not fit the installed Falco version. Currently, it happens before trying a local prebuilt and the download strategy, but that should be the last resort.

**Comment by leogr** [comment]
Rebased on top of master to include the fix introduced by https://github.com/falcosecurity/falco/pull/1212

I removed WIP, because it's ok for me, though it's not possible to test every single case.

@fntlnz @leodido So, question: do we want this PR to be merged for the 0.23.0?

**Comment by leogr** [comment]
Rebased on top of master to include lastest fixes.

**Pull request merged**